drivers/soundwire/intel_ace2x.c

```diff
@@ -34,37 +36,56 @@ static void intel_shim_vs_init(struct sdw_intel *sdw)
 
 static int intel_shim_check_wake(struct sdw_intel *sdw)
 {
-	void __iomem *shim_vs;
+	u16 lsdiid = 0;
 	u16 wake_sts;
+	int ret;
+
+	/* find out which bits are set in LSDIID for this sublink */
+	ret = hdac_bus_eml_sdw_get_lsdiid_unlocked(sdw->link_res->hbus, sdw->instance, &lsdiid);
+	if (ret < 0)
+		return ret;
 
-	shim_vs = sdw->link_res->shim_vs;
-	wake_sts = intel_readw(shim_vs, SDW_SHIM2_INTEL_VS_WAKESTS);
+	/*
+	 * we need to use the global HDaudio WAKEEN/STS to be able to detect
+	 * wakes in low-power modes
+	 */
+	wake_sts = snd_hdac_chip_readw(sdw->link_res->hbus, STATESTS);
 
-	return wake_sts & SDW_SHIM2_INTEL_VS_WAKEEN_PWS;
+	return wake_sts & lsdiid;
 }
 
 static void intel_shim_wake(struct sdw_intel *sdw, bool wake_enable)
 {
-	void __iomem *shim_vs = sdw->link_res->shim_vs;
+	u16 lsdiid = 0;
 	u16 wake_en;
 	u16 wake_sts;
+	int ret;
+
+	mutex_lock(sdw->link_res->shim_lock);
```

Comment:
shim_lock is used to protect sdw shim register access, but here we only access hda bus. should we use hbus->lock ?

Reply:
good feedback but this is NOT only to protect the shim registers, which no longer exist in LNL. See https://github.com/thesofproject/linux/blob/topic/sof-dev/drivers/soundwire/intel_init.c#L78, this 'shim_lock' is actually the eml_lock.